### ISSUE TITLE
ECH contradicts second ClientHello consistency requirements in RFC8446

### ISSUE THREAD

**Issue by davidben** [tlswg/draft-ietf-tls-esni#358]
I just noticed this as I was pondering our various HRR issues. RFC8446, section 4.1.2 says:

> [...] The client will also send a ClientHello when the server has responded to its ClientHello with a HelloRetryRequest.  In that case, the client MUST send the same ClientHello without modification, except as follows:
>
> * [...various cases regarding existing extensions...]
>
> *  Other modifications that may be allowed by an extension defined in the future *and present in the HelloRetryRequest*.

https://tools.ietf.org/html/rfc8446#section-4.1.2

The "and present in the HelloRetryRequest" portion is fun. In RFC8446 as written, we're not allowed to change the `encrypted_client_hello` extension on the second ClientHello unless the server sent `encrypted_client_hello` in HelloRetryRequest, which it doesn't. With the protocol as-is, it seems we'll at least need some text to deal with the contradiction.

The impact is practice is thankfully limited. If the servers does not enforce consistency between the two ClientHellos, it doesn't matter. (They're not required to, but they're not explicitly forbidden from it either, and a stateless server may need to enforce some minimum consistency to avoid getting into a broken state.) If a server does, ECH's behavior on HRR will break it. Such a server is supposed to handshake with ClientHelloOuter, so it just affects the retry flow. That's maybe okay, but mildly annoying. Administrators using server software that checks need to know to first upgrade to a version that doesn't check before trying to deploy ECH.

We can also add an ECH extension to HRR. Then, on ECH-reject + HRR, the server perhaps wouldn't send the extension and the client would be required to replay the extension. This is a waste in many ways, but would work. This would break don't stick out, but some threat models, don't stick out with HRR is hopeless anyway. (See https://github.com/tlswg/draft-ietf-tls-esni/pull/352#issuecomment-721824226.) It would also be weirdly asymmetric between ECH and SH.

That said, this constraint is also pretty obnoxious. For instance, #313 currently adds a new padding extension codepoint, because the old one isn't defined for other messages. However, the RFC8446 allowance only works for the existing padding code point, so you don't want to send any other code points in ClientHello, lest the server force you to keep it unchanged in ClientHello2. So maybe we should confirm no servers did the fully strict check and undo that constraint? (Although, for #313, I personally favor the padding message strategy anyway.)

**Comment by davidben** [comment]
@cjpatton points out the Go implementation does check CH1 and CH2 consistency. Fortunately, it only checks known extensions, so it's still compatible with the ECH draft.
https://golang.org/src/crypto/tls/handshake_server_tls13.go#L457

Checking only know extensions is basically saying we should strike "and present in the HelloRetryRequest" in rfc8446bis and make sure everyone knows not to check this.

**Comment by ekr** [comment]
I don't think we can *just* strike this text because the point of that text was that you couldn't like change your KeyShares if the HRR didn't tell you to. But I agree we can sharpen this in some sensible way.

**Comment by davidben** [comment]
I figured that case was covered by "defined in the future", which I'm not proposing we delete. key_share isn't defined in the future, so it falls under existing rules. The existing rules (bullet points over that one) only let you change key_share "[i]f a "key_share" extension was supplied in the HelloRetryRequest". But perhaps that's a little roundabout. The rules I'm thinking of are (changes *emphasized*):

* Each extension gets to define how it reacts to HRR.
* It typically is sensitive the corresponding HRR extension, *but doesn't have to be*.
* RFC8446(bis) defines the rules for all existing extensions.
* Future extensions define their own rules. (And probably if they don't say anything, we assume they don't get to change.)
* As a server, you can make assumptions based on the rules for extensions you know about. *You have to tolerate random changes in extensions you don't know about, because you don't know their rules.*

I'm running a probe on a bunch of servers to see how they react to changes in cipher suites, curves, and a random extension. It seems most don't notice. Some (I'm guessing Go) notice cipher suites and curves. None notice a random extension. That suggests the change will be compatible.

**Comment by davidben** [comment]
> I'm running a probe on a bunch of servers to see how they react to changes in cipher suites, curves, and a random extension. It seems most don't notice. Some (I'm guessing Go) notice cipher suites and curves. None notice a random extension. That suggests the change will be compatible.

I spoke too soon. LibreSSL breaks if you change an extension they don't recognize. :-(

**Comment by davidben** [comment]
So draft-09 isn't tied up with this, I've uploaded https://github.com/tlswg/draft-ietf-tls-esni/pull/361 as a bandaid for the potential compatibility issue.

(I say potential because it's only a problem if the server checks unknown extensions *and* has different enough key share preferences from the client that it needs HRR. HRR is pretty rare, so it may not actually be an issue. For Chrome, at least, I don't think the particular servers I found actually trigger HRR. But it is still a risk for clients or servers with different configurations.)
